### PR TITLE
Finalizing Batch #5b tests

### PR DESCRIPTION
**Pull request by Asmaloth**
## Related Issue
No Jira ticket currently exists but for more documentation visit: https://www.notion.so/webiny/Webiny-Testing-43be8707c3df48cbbf87d06a57720faa

## Your solution
Using a combination of UI tests and graphQL commands we have been able to achieve the wanted results.

## How has this been tested?
Using eslint and CI checks.

## Screenshots (if relevant):
N/A

### CODE REVIEW

File: cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    const menuNameEdit = "Testing Menu123";
+    const menuDescEdit = "This is an edited description.";
+
+    beforeEach(() => {
+        cy.login();
+        cy.pbDeleteAllMenus();
+    });
+
+    afterEach(() => {
```
Comment (adrians5j) [inline cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts]: After each here? 🤔

File: cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        cy.login();
+        cy.pbDeleteAllMenus();
+    });
+    it("should be able to create, edit, and immediately delete a menu", () => {
+        cy.visit("/page-builder/menus");
+        cy.wait(1000);
+        //Create a menu.
```
Comment (adrians5j) [inline cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts]:
```suggestion
        // Create a menu.
```

File: cypress-tests/cypress/support/pageBuilder/reloadUntil.ts
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/reloadUntil.ts:1]: We already have this, why is this file added? 🤔

File: cypress-tests/cypress/support/pageBuilder/pbUpdatePage.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                }
+                error {
+                    code
+                    message
+                    data
+                    __typename
+                }
+                __typename
+            }
+            __typename
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbUpdatePage.ts]: No need to include `__typename` in GQL queries / mutations.

File: cypress-tests/cypress/support/pageBuilder/pbUpdatePage.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                variables: data,
+                authToken: user.idToken.jwtToken
+            })
+            .then(response => {
+                const { error, data } = response.pageBuilder.updatePage;
+                if (error) {
+                    throw new Error(`Failed to update page: ${error.message}`);
+                }
+
+                return cy.wrap(data); // Wrap the data in a Chainable
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbUpdatePage.ts]: Curious about this wrap here. Can you explain?

File: cypress-tests/cypress/support/pageBuilder/pbPublishPage.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                    __typename
+                }
+                __typename
+            }
+            __typename
+        }
+    }
+`;
+
+// Update the pbPublishPage method
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbPublishPage.ts]: Is this comment relevant?

File: cypress-tests/cypress/support/pageBuilder/pbListPages.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+declare global {
+    // eslint-disable-next-line @typescript-eslint/no-namespace
+    namespace Cypress {
+        interface Chainable {
+            pbListPages(data: any): Chainable<Promise<any[]>>;
+        }
+    }
+}
+
+const ERROR_FIELDS = `
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbListPages.ts]: Feel free to paste these fields directly into L35, without using `ERROR_FIELDS`. Remove `ERROR_FIELDS` afterwards.

File: cypress-tests/cypress/support/pageBuilder/pbDeletePage.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import { gqlClient } from "../utils";
+
+declare global {
+    // eslint-disable-next-line @typescript-eslint/no-namespace
+    namespace Cypress {
+        interface Chainable {
+            pbDeletePage(data: any): Chainable<Promise<[]>>;
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbDeletePage.ts]:
```suggestion
            pbDeletePage(params: { id: string }): Chainable<Promise<any>>;
```

File: cypress-tests/cypress/support/pageBuilder/pbDeleteAllMenus.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                data {
+                    title
+                    slug
+                    description
+                    items
+                    createdOn
+                    createdBy {
+                        id
+                        displayName
+                        __typename
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbDeleteAllMenus.ts]: Just in case - make sure `__typename` is removed on all of these queries that this PR touched.

**Review by adrians5j — CHANGES_REQUESTED** [review]
Did another iteration, but I sense there will be more to double check, especially around types.

File: cypress-tests/cypress/support/pageBuilder/reloadUntil.ts
Comment (Asmaloth) [reply cypress-tests/cypress/support/pageBuilder/reloadUntil.ts:1]: I though I'd be better to have it be named reloadUntil.ts rather than index.js
image
But I did not want to delete the original file as it's still being used elsewhere. I might need some consulting regarding this but in a lot of places in this PR I tried to make a .ts equivalent for a .js file I have probably failed somewhere as well

File: cypress-tests/cypress/support/pageBuilder/pbUpdatePage.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+                    status
+                    savedOn
+                }
+            }
+        }
+    }
+`;
+
+Cypress.Commands.add("pbUpdatePage", data => {
+    cy.login().then(user => {
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbUpdatePage.ts]:
```suggestion
    return cy.login().then(user => {
```

File: cypress-tests/cypress/support/pageBuilder/pbUpdatePage.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            };
+        };
+    };
+}
+
+declare global {
+    // eslint-disable-next-line @typescript-eslint/no-namespace
+    namespace Cypress {
+        interface Chainable {
+            pbUpdatePage(data: PbUpdatePageInput): Promise<string>;
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbUpdatePage.ts]: In L58, we're returning `data` which is an object, not string.

Let's switch `string` in `Promise<string>` to an object that consists of keys defined in the GQL query, starting L32.

File: cypress-tests/cypress/support/pageBuilder/pbPublishPage.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+}
+
+const PUBLISH_PAGE = /* GraphQL */ `
+    mutation PbPublishPage($id: ID!) {
+        pageBuilder {
+            publishPage(id: $id) {
+                data {
+                    id
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbPublishPage.ts:8]: Let's remove redundant fields here and sync what's left with the return type in L7.

File: cypress-tests/cypress/support/pageBuilder/pbPublishPage.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+                        version
+                    }
+                }
+            }
+        }
+    }
+`;
+
+Cypress.Commands.add("pbPublishPage", id => {
+    cy.login().then(user => {
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbPublishPage.ts]:
```suggestion
    return cy.login().then(user => {
```

File: cypress-tests/cypress/support/pageBuilder/pbListPages.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+                    data
+                    message
+                }
+            }
+        }
+    }
+`;
+
+Cypress.Commands.add("pbListPages", data => {
+    cy.login().then(user => {
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbListPages.ts]:
```suggestion
    return cy.login().then(user => {
```

File: cypress-tests/cypress/support/pageBuilder/pbListPages.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
+import { gqlClient } from "../utils";
+
+declare global {
+    // eslint-disable-next-line @typescript-eslint/no-namespace
+    namespace Cypress {
+        interface Chainable {
+            pbListPages(data: any): Chainable<Promise<any[]>>;
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbListPages.ts]:
```suggestion
            pbListPages(data: any):Promise<Array<{GQL FIELDS YOU HAVE IN LIST_PAGES QUERY}>>;
```

File: cypress-tests/cypress/support/pageBuilder/pbDeletePage.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+                    data
+                    code
+                }
+            }
+        }
+    }
+`;
+
+Cypress.Commands.add("pbDeletePage", data => {
+    cy.login().then(user => {
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbDeletePage.ts]:
```suggestion
   return cy.login().then(user => {
```

File: cypress-tests/cypress/support/pageBuilder/pbDeleteAllMenus.ts [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    namespace Cypress {
+        interface Chainable {
+            listMenus(): Promise<Record<string, any>[]>;
+            deleteMenu(slug: string): Promise<Record<string, any>>;
+            pbDeleteAllMenus(): void;
+        }
+    }
+}
+
+Cypress.Commands.add("listMenus", () => {
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbDeleteAllMenus.ts:39]: If we want to have these commands potentially for later use (they are not being used just in this file), let's extract these new commands into separate files, the same way we're doing with other commands.

P.S. Don't forget the `ps` prefix, for example `pbListMenus`.

File: cypress-tests/cypress/support/pageBuilder/pbDeleteAllMenus.ts [hunk 17]
```diff
@@ -0,0 +1,82 @@
+import { gqlClient } from "../utils";
+
+const LIST_MENUS_QUERY = /* GraphQL */ `
+    query listMenus {
+        pageBuilder {
+            listMenus {
+                data {
+                    title
+                    slug
+                    description
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbDeleteAllMenus.ts]: Check which fields we actually need, and we don't need.

File: cypress-tests/cypress/support/pageBuilder/pbCreatePage.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+
+const CREATE_PAGE = /* GraphQL */ `
+    mutation PbCreatePage($from: ID, $category: String, $meta: JSON) {
+        pageBuilder {
+            createPage(from: $from, category: $category, meta: $meta) {
+                data {
+                    id
+                    pid
+                    status
+                    title
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbCreatePage.ts:9]: Check which fields we need and which we don't need.

File: cypress-tests/cypress/support/pageBuilder/pbCreatePage.ts [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+                }
+            }
+        }
+    }
+`;
+declare global {
+    // eslint-disable-next-line @typescript-eslint/no-namespace
+    namespace Cypress {
+        interface Chainable {
+            pbCreatePage(data: any): Promise<any>;
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbCreatePage.ts]:
```suggestion
            pbCreatePage(data: any): Promise<{FIELDS_YOU_INCLUDE_IN_GQL_OPERATION}>;
```

File: cypress-tests/cypress/support/pageBuilder/pbClearMainMenu.ts [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+            pbClearMainMenu(): Chainable<Promise<any>>;
+        }
+    }
+}
+
+// GraphQL mutation to clear menu items
+const CLEAR_MAIN_MENU = `
+  mutation updateMenu($slug: String!, $data: PbMenuInput!) {
+    pageBuilder {
+      menu: updateMenu(slug: $slug, data: $data) {
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/pbClearMainMenu.ts:20]: Check the same things reported in other commands.

File: cypress-tests/cypress/support/pageBuilder/createMenu.ts [hunk 21]
```diff
@@ -0,0 +1,45 @@
+import { gqlClient } from "../utils";
+
+declare global {
+    // eslint-disable-next-line @typescript-eslint/no-namespace
+    namespace Cypress {
+        interface Chainable {
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/createMenu.ts]: Check the same things reported in other commands.

File: cypress-tests/cypress/e2e/admin/pageBuilder/menus/pagesListMenuItemType.cy.js
Comment (adrians5j) [inline cypress-tests/cypress/e2e/admin/pageBuilder/menus/pagesListMenuItemType.cy.js:1]: We have both js and ts versions of this file.

File: cypress-tests/cypress/support/pageBuilder/reloadUntil.ts [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+const MAX_RETRIES = 100;
+const WAIT_BETWEEN_RETRIES = 3000;
+const REPEAT_WAIT_BETWEEN_RETRIES = 2000;
+
+const sleep = (ms = WAIT_BETWEEN_RETRIES) =>
+    new Promise(resolve => {
+        setTimeout(resolve, ms);
+    });
+
+Cypress.Commands.add("reloadUntil", (callback, options = {}) => {
```
Comment (adrians5j) [inline cypress-tests/cypress/support/pageBuilder/reloadUntil.ts:20]: We're missing TS declaration here.

File: cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+
+        cy.findByTestId("pb.menu.create.name").type(menuName);
+        cy.findByTestId("pb.menu.save.button").click();
+        cy.findByText("Value is required.").should("exist");
+        cy.findByTestId("pb.menu.create.slug").type(menuSlug);
+        cy.findByTestId("pb.menu.create.description").type("This is a cool test.");
+        cy.findByTestId("pb.menu.save.button").click();
+
+        cy.findByText("Menu saved successfully.").should("exist");
+        cy.wait(500);
```
Comment (adrians5j) [inline cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts]: Let's not use cy.wait if really not needed. If needed, let's ensure a comment explains what it's there.

Let's adopt the same approach in all future tests.

File: cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+
+        // Assert the menu is being displayed properly on the right side of the screen.
+        cy.findByTestId("default-data-list").within(() => {
+            cy.get("li")
+                .first()
+                .within(() => {
+                    cy.findByText(menuNameEdit).should("exist");
+                    cy.findByText(menuDescEdit).should("exist");
+                    cy.get("button").click({ force: true });
```
Comment (adrians5j) [inline cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts]: Let's use a test ID here. 

The moment you're forced to use a vague selector like this one, just ping me and I'll add a test ID.

File: cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+                });
+        });
+
+        //Finish deleting the menu and assert it is deleted.
```
Comment (adrians5j) [inline cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts]:
```suggestion
        // Finish deleting the menu and assert it is deleted.
```

File: cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts [hunk 26]
```diff
@@ -0,0 +1,59 @@
+context("Page Builder - Menu CRUD", () => {
+    const menuName = "Test menu";
```
Comment (adrians5j) [inline cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts]: Can we use nanoid here? Just so we're consistent.

**Review by adrians5j — CHANGES_REQUESTED** [review]
Mostly, some code-level things to take a look at. Let's get this in order first, just so we don't end up with a pile of bad code in the near future. :)

File: cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts (hunk 24 above)
Comment (adrians5j) [reply cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts]: Will add a test ID now here... hold on...

Comment (adrians5j) [reply cypress-tests/cypress/e2e/admin/pageBuilder/menus/menuCrud.cy.ts]: image

```
pb-menus-list-delete-menu-btn
```

File: cypress-tests/cypress/support/pageBuilder/pbListPages.ts (hunk 13 above)
Comment (Asmaloth) [reply cypress-tests/cypress/support/pageBuilder/pbListPages.ts]: Adding return here causes other issues. May we discuss this further tomorrow?

File: cypress-tests/cypress/support/pageBuilder/pbDeletePage.ts (hunk 15 above)
Comment (Asmaloth) [reply cypress-tests/cypress/support/pageBuilder/pbDeletePage.ts]: Adding return here also causes otther issues

File: cypress-tests/cypress/support/pageBuilder/pbDeleteAllMenus.ts (hunk 16 above)
Comment (Asmaloth) [reply cypress-tests/cypress/support/pageBuilder/pbDeleteAllMenus.ts:39]: I have separated them into pbListMenus and pbDeleteMenu with pbDeleteAllMenus remaining as well.

Sadly, none of it works anymore when I separate them and i'm really not sure why. I tried debugging cy.pbListPages but it always returns a null. We can check it out in our next call if we get the time.

File: cypress-tests/cypress/support/pageBuilder/pbDeleteAllMenus.ts (hunk 17 above)
Comment (Asmaloth) [reply cypress-tests/cypress/support/pageBuilder/pbDeleteAllMenus.ts]: This is being looked into as part of the comments above.

File: cypress-tests/cypress/support/pageBuilder/createMenu.ts (hunk 21 above)
Comment (Asmaloth) [reply cypress-tests/cypress/support/pageBuilder/createMenu.ts]: I am having many issues getting this up to par with the other commands.
Since the old command was working properly I have deleted this filed and moved the old pbCreateMenu.js -> pbCreateMenu.ts

File: cypress-tests/cypress/e2e/admin/pageBuilder/menus/pagesListMenuItemType.cy.js
Comment (Asmaloth) [reply cypress-tests/cypress/e2e/admin/pageBuilder/menus/pagesListMenuItemType.cy.js:1]: I did not want to remove .js files until we are completely done with the .ts files. Would you like me to do this know or will we cleanup all the batches of remaining .js files at once.

File: cypress-tests/cypress/support/pageBuilder/pbDeletePage.ts (hunk 15 above)
Comment (Asmaloth) [reply cypress-tests/cypress/support/pageBuilder/pbDeletePage.ts]: Issue resolved by removing Chainable from line 7

**Comment by adrians5j** [comment]
No longer needed, merged in a separate PR.
